KEP-3619: Fine-grained SupplementalGroups control

everpeace opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Implements https://github.com/kubernetes/enhancements/issues/3619

#### Special notes for your reviewer:

The actual policy enforcement will be implemented by container runtimes (cri-api implementations). Because most container runtimes (e.g. containerd) depend on https://github.com/kubernetes/cri-api, this PR should be merged before container runtimes implement actual policy enforcement. Please also see [CRI API | Feature development](https://github.com/kubernetes/kubernetes/blob/6546c8b/staging/src/k8s.io/cri-api/README.md#feature-development).


My PR in containerd is already opened and in review: https://github.com/containerd/containerd/pull/9737. It already got LGTM (https://github.com/containerd/containerd/pull/9737#pullrequestreview-2057591573) except for cri-api dependency, which will be updated once this PR gets merged and v1.31 alpha or beta version is released.


_If you wanted to try running this feature, you can try it with kind easily: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39_

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Support fine-grained supplemental groups policy (KEP-3619), which enables fine-grained control for supplementary groups in the first container processes. You can choose whether to include groups defined in the container image(/etc/groups) for the container's primary uid or not.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3619-supplemental-groups-policy
```

everpeace committed:
Add `Linux{Sandbox,Container}SecurityContext.SupplementalGroupsPolicy` and `ContainerStatus.user` in cri-api

everpeace committed:
Implement kubelet so to wire between SecurityContext.SupplementalGroupsPolicy/ContainerStatus.User and cri-api in kubelet

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test all

everpeace commented:
/test pull-kubernetes-e2e-capz-windows

everpeace commented:
/cc  @thockin @mrunalp @SergeyKanzhelev

Let me add you guys as [this KEP reviewers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3619-supplemental-groups-policy/kep.yaml#L10C1-L15)

cici37 commented:
/remove-sig api-machinery

everpeace commented:
rebased.

everpeace commented:
/remove-sig api-machinery

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -18,6 +18,7 @@ package validation
 
 import (
 	"fmt"
+
```
can you just drop this file?

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2505,6 +2505,32 @@ type ContainerStatus struct {
+type ContainerUser struct {
+	// Linux holds user identity information of the first process of the containers in Linux.
+	// Note that this field cannot be set when spec.os.name is windows.
+	Linux *LinuxContainerUser
```
I don't love spelling out Linux here - is there precedent?

everpeace replied:
@thockin sorry for long delay.

> is there precedent?

As far as I know, probably `seLinuxOptions` in `SecurityContext`?

However, I'm not a Windows specialist, though, UID/GID is a specific concept in Linux, isn't it? Then, I don't have a good idea to express this information in the container status.

Would you have suggestions?

everpeace replied:
@thockin I keep this as is because I still have no idea how to avoid spelling out Linux here because supplementalGroups is the idea which is specific to Linux🙇 Any suggestions??

thockin replied:
@liggitt @tallclair for consult.  I hate that the API spells out "Linux" but the structure would be different for Windows.

liggitt replied:
It is ugly, but I agree the shape for linux and windows is usually different.

For prior art, we have `securityContext.windowsOptions` at the pod and container level. In the cri-api, we also have 

```
    // Optional configurations specific to Linux hosts.
    LinuxPodSandboxConfig linux = 8;
    // Optional configurations specific to Windows hosts.
    WindowsPodSandboxConfig windows = 9;
```

For OS-specific fields like this, do validate that this is not set if `spec.podOS.name` is set to something other than "linux"

thockin replied:
ACK

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4421,6 +4424,24 @@ func validateSysctls(sysctls []core.Sysctl, fldPath *field.Path) field.ErrorList
+func validateSupplementalGroupsPolicy(supplementalGroupsPolicy *core.SupplementalGroupsPolicy, podOs *core.PodOS, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
This is not how we usually gate new fields.

The gate controls admission into the system.  Once it is in, we use it.  What that means practically is that you want to add this in k8s.io/kubernetes/pkg/api/pod `DropDisabledPodFields()` (see the other examples).

Then, in here, you do something like:

```
if gate enabled {
    if field is nil {
        field.Required(...)
    } else if field is not a valid value {
            field.Invalid(...)
    }
}
```

You also need to make sure that no updates are allowed when the gate is disabled.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4775,6 +4798,15 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 func ValidateContainerStateTransition(newStatuses, oldStatuses []core.ContainerStatus, fldpath *field.Path, restartPolicy core.RestartPolicy) field.ErrorList {
 	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
same as above, except this is not required

thockin replied:
Actually, I am not sure there's any validation you CAN do here.  Maybe check thart UID and GIDs are not negative?

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2505,6 +2505,32 @@ type ContainerStatus struct {
+// LinuxContainerUser represents user identity information in Linux containers
+type LinuxContainerUser struct {
+	// Uid is the primary uid of the first process in the container
+	Uid int64
```
UID and GID

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2723,6 +2723,34 @@ type ContainerStatus struct {
+// LinuxContainerUser represents user identity information in Linux containers
+type LinuxContainerUser struct {
+	// Uid is the primary uid of the first process in the container
+	Uid int64 `json:"uid,omitempty" protobuf:"varint,1,name=uid"`
```
don't omitempty - we want 0 to print

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2723,6 +2723,34 @@ type ContainerStatus struct {
+	// Uid is the primary uid of the first process in the container
+	Uid int64 `json:"uid,omitempty" protobuf:"varint,1,name=uid"`
+	// Gid is the primary gid of the first process in the container
+	Gid int64 `json:"gid,omitempty" protobuf:"varint,2,name=gid"`
```
same

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3788,9 +3836,17 @@ type PodSecurityContext struct {
+	// (Alpha) Using the field requires the SupplementalGroupsPolicy feature gate to be enabled
+	// and the container runtime must implement support for this feature.
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
```
I think you can `+default=Merge` here and avoid manual defaulting

everpeace replied:
I think `SupplementalGroupsPolicy` is os dependent. We should reject this field when os is windows. So, I didn't put the `+default` tag and keeps windows validation.

https://github.com/kubernetes/kubernetes/pull/117842/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR4283-R4285

thockin reviewed (commented):
You need to set the default value in pkg/apis/core/v1/defaults.go.

thockin commented:
I am reviewing PRs today and Monday, and then I am out - if we want to get this into 1.28, it needs to be updated ASAP.

Atharva-Shinde commented:
/milestone v1.28

Atharva-Shinde commented:
/milestone clear

thockin commented:
I see this got missed - is it still something we want?  I think it's a good change, we just need to revisit the PR....

everpeace commented:
> I think it's a good change, we just need to revisit the PR....

Thanks, I agree.

---

Sorry for my long absence. Let me revisit this PR in the next week.

This PR includes:
- CRI changes
- k8s API changes
- Implementation
  - k8s api validation
  - wiring SupplementalGroupPolicy/LinuxUser to CRI
  - bubbling up LinuxContainerUser from CRI to k8s api

However, no container runtime does support this feature currently. So, I'm thinking I should split up this PR into several PRs

- PR 1: CRI change
- PR 2: implementation the feature in containerd or CRI-O 
- PR 3: k8s API changes and implementation

What do you think??

thockin commented:
The problem with breaking a PR up like that is that it is SO EASY to get stuck in the middle with something that can't be used but also can't easily be reverted.  I'd rather it go in all at once, even if there's no runtime support yet, and stay alpha until "enough" runtimes handle it

thockin commented:
Also, we might hold off on merging it until at least one major runtime supports it, so we can ACTUALLY prove it.

everpeace commented:
I see, I understood your point. And, I also found this document: https://github.com/kubernetes/cri-api?tab=readme-ov-file#feature-development

This document says:
> The recommended feature development flow is following::
> - Review proposed CRI API changes during the KEP review stage. Some field names and types may not be spelled out exactly at this stage.
> - Locally implement a prototype that implement changes in both - Kubernetes and Container Runtime.
> - Submit a Pull Request for Kubernetes implementing CRI API changes alongside the feature code. Feature must be developed to degrade gracefully when used with older Container Runtime according to the Version Skew policy.

Let me follow this path. I will keep working on this PR.

thockin commented:
Pinging this old one, which I still would like to see happen.

everpeace commented:
Sorry for my absence the long time. Let me resume my work. I will update my branch soon.

everpeace commented:
I merged master branch for ease of review. I will rebase to the latest master once review is approved.

If it is better to rebase, I would be happy if you tell me so 🙇

everpeace commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-capz-windows-master

everpeace commented:
/retest pull-kubernetes-conformance-kind-ga-only-parallel

everpeace commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

everpeace reviewed (commented):
~~@thockin @mrunalp @SergeyKanzhelev Sorry for my long absence.~~

~~I updated my branch(just merged master for smooth review) and dropped my old validation logic and implemented `DropDisabledPodFields` as @thockin [pointed out](https://github.com/kubernetes/kubernetes/pull/117842/files/2808dde105103b08498a687b5d3fbdaf67f6dff8#r1210928751). I'm also working on cri implementation in https://github.com/containerd/containerd/pull/9737~~

~~You can try this feature with kind: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39#file-readme-md~~

~~Please take a look 🙇~~

~~Note:~~
~~- Once got final approval, I will rebase to the latest master~~

everpeace committed:
Add `PodSecurityContext.SupplementalGroupsPolicy`, `ContainerStatus.User` and its featuregate

everpeace commented:
/remove-lifecycle stale

everpeace reviewed (commented):
@thockin @mrunalp @SergeyKanzhelev (KEP reviewers)

Hi, let me reboot this PR. I would like to aim for this feature to be released in v1.31. 

I rebased my branch to the latest master because my branch was too stale to merge back to it. Could you review this PR again🙇🙇? 

As you know, this KEP requires implementing CRI on the container runtime side. I already opened a PR https://github.com/containerd/containerd/pull/9737 as the first supported runtime for alpha release [as described in KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3619-supplemental-groups-policy#alpha). 

_Added: the https://github.com/containerd/containerd/pull/9737 got LGTM(https://github.com/containerd/containerd/pull/9737#pullrequestreview-2057591573) except for cri-api dependency._

_If you wanted to try running this feature, you can try it with kind easily: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39_

ktock commented on pkg/api/pod/util.go:
```diff
@@ -809,16 +818,16 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
 		podStatus.HostIPs = nil
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) && !rroInUse(oldPodSpec) {
-		for i := range podStatus.ContainerStatuses {
-			podStatus.ContainerStatuses[i].VolumeMounts = nil
-		}
-		for i := range podStatus.InitContainerStatuses {
-			podStatus.InitContainerStatuses[i].VolumeMounts = nil
-		}
-		for i := range podStatus.EphemeralContainerStatuses {
-			podStatus.EphemeralContainerStatuses[i].VolumeMounts = nil
```
Is this change related to this PR?

everpeace replied:
Thanks for the catch! I put this code back🙇

everpeace replied:
addressed in 2db0c01. sorry for the rebase 🙇

everpeace committed:
Implement DropDisabledPodFields for PodSecurityContext.SupplementalGroupsPolicy and ContainerStatus.User fields

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3629,9 +3673,17 @@ type PodSecurityContext struct {
 	// +optional
 	SupplementalGroups []int64
+	// Defines how supplemental groups of the first container processes are calculated.
+	// Valid values are "Merge" and "Strict". If not specified, "Merge" is used.
```
Previously we said we cannot have a literal default value because of bugs, but once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly.

thockin replied:
The bug today is that a new field with a default value gets considered by kubelet as a change, so it restarts all pods (and all nodes do this roughly in sync).  #124220  fixes that, so a new default should not trigger a restart.

> this field should be an OS dependent field

OK, thast may still be true, but adding a default will avoid code like:

```
if windows {
  return doSomethingElse()
}
policy = merge // hardcoded default
if FSGroupChangePolicy == nil {
    policy = *FSGroupChangePolicy
}
return doSomethingWithPolicy(policy)
```

And when users `kubectl get` the pods they will see "Merge" instead of nothing

everpeace replied:
> so a new default should not trigger a restart

Wow, Thanks. That is why we have such constraint in `default_test.go`, right?

https://github.com/kubernetes/kubernetes/blob/5f1e5cafd33a8863c5dfe9f36c9cba836f0e981b/pkg/apis/core/v1/defaults_test.go#L165-L174

I didn’t know that. I’ve learned something today.

<hr />

> but adding a default will avoid code like:
> And when users kubectl get the pods they will see "Merge" instead of nothing

Thanks for this, too. Probably, now I got your idea. Once #124220 got landed, we can introduce to add new field with default value. But, this time, we would like to introduce conditional(os specific) default value like this:

- when `spec.os.name == windows`, it should validate the field must be `nil` 
- when `spec.os.name == linux`, it should make the default value be `Merge`

Then, perhaps, we can use custom defaulter function instead of `+default=“Merge”` annotation to achieve conditional default value?? I imagine like below:

```golang
# in pkg/apis/core/v1/defaults.go
func SetDefaults_PodSpec(obj *v1.PodSpec) {
…

	if obj.SecurityContext == nil {
		obj.SecurityContext = &v1.PodSecurityContext{}
	}
	if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
		if obj.SecurityContext.SupplementalGroupsPolicy == nil && (obj.OS == nil || obj.OS.Name == v1.Linux) {
			merge := v1.SupplementalGroupsPolicyMerge
			obj.SecurityContext.SupplementalGroupsPolicy = &merge
		}
	}
…
```

thockin replied:
@alexzielenski Here's a fun case for declarative defaults  :)

@everpeace - something like that, yes.

liggitt replied:
> once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly

well... once it lands and makes it to the oldest kubelet that could be running against the server (so once the n-3 kubelet is 1.31, so 1.34)

thockin replied:
oh, yeah, I guess that's a good point.  I got a little ahead of myself.  Sorry @everpeace - what we CAN do is leave "TODO" comments to revisit this in 1.34 :)

thockin commented:
I am still keen to see this go in!  Thanks for keeping it going @everpeace !

everpeace committed:
Clarify `SupplementalGroupsPolicy` is an OS depdendent field.

everpeace committed:
Make `ContainerStatus.User` is initially attached user identity to the first process in the ContainerStatus

It is because, the process identity can be dynamic if the initially attached identity
has enough privilege calling setuid/setgid/setgroups syscalls in Linux.

everpeace commented on pkg/apis/core/types.go:
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
 	// +optional
 	// +featureGate=RecursiveReadOnlyMounts
 	VolumeMounts []VolumeMountStatus
+	// User represents user identitiy information initially attached to the first process of the container
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
+	User *ContainerUser
+}
+
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual process identity can be dynamic if the initially attached identity have enough privilege calling setuid/setgid/setgroups syscalls
+	// +optional
+	Linux *LinuxContainerUser
```
I emphasize the information is __initially attached__ one here.

As @ktock pointed out in https://github.com/containerd/containerd/pull/9737#discussion_r1598603468, process identity can be dynamic/updated if the initially attached identity has enough privilege calling setuid/setgid/setgroups syscall.

everpeace commented:
/test pull-kubernetes-e2e-capz-windows-master

thockin commented:
This needs sig-node review and approval:  @SergeyKanzhelev @mrunalp @dchen1107 ?

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual process identity can be dynamic if the initially attached identity have enough privilege calling setuid/setgid/setgroups syscalls
```
Propose a reword:

```
Note that the actual running identity can be changed if the process has enough privilege to do so.
```

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3541,6 +3568,24 @@ const (
+type SupplementalGroupsPolicy string
+
+const (
+	// Merge policy always merges the provided SupplementalGroups (including FsGroup)
```
Wording:

```
	// SupplementalGroupsPolicyMerge means that the container's provided
	// SupplementalGroups and FsGroup (specified in SecurityContext) will be
	// merged with the primary user's groups as defined in the container image
	// (in /etc/group).
```

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3541,6 +3568,24 @@ const (
+	// Merge policy always merges the provided SupplementalGroups (including FsGroup)
+	// specified in SecurityContext with groups of the primary user from the container
+	// image(`/etc/group`).
+	// Note: The primary user is specified with RunAsUser.
```
I think we can drop this part, or put it on the RunAsUser field

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3541,6 +3568,24 @@ const (
+	//       If not specified, the user from the image config is used.
+	//       Otherwise, the runtime default is used.
+	SupplementalGroupsPolicyMerge SupplementalGroupsPolicy = "Merge"
+	// Strict policy uses only the provided SupplementalGroups(including FsGroup)
```
Wording

```
	// SupplementalGroupsPolicyStrict means that the container's provided
	// SupplementalGroups and FsGroup (specified in SecurityContext) will be
	// used instead of any groups defined in the container image.
```

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3629,9 +3674,18 @@ type PodSecurityContext struct {
 	// no additional groups are added to any container. Note that group memberships
 	// defined in the container image for the uid of the container process are still effective,
 	// even if they are not included in this list.
+	// If SupplementalGroupsPolicy feature was enabled, you can control how the groups are calculated.
```
Can we reword this wole comment?

```
	// A list of groups applied to the first process run in each container, in
	// addition to the container's primary GID and fsGroup (if specified).  If
	// the SupplementalGroupsPolicy feature is enabled, the
	// supplementalGroupsPolicy field determines whether these are in addition
	// to or instead of any group memberships defined in the container image.
	// If unspecified, no additional groups are added, though group memberships
	// defined in the container image may still be used, depending on the
	// supplementalGroupsPolicy field.
```

everpeace committed:
Rewording suggestion applied

everpeace commented:
/test pull-kubernetes-integration

everpeace reviewed (commented):
Thank you for re-wording suggestions. Addressed in below packages:
- `pkg/core/v1`
- `staging/k8s.io/api/core/v1`
- `staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto`

everpeace committed:
Add TODO comment for updating SupplementalGroupsPolicy default value in v1.34

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual running identity can be changed if the process has enough privilege to do so.
+	// +optional
+	Linux *LinuxContainerUser
```
I think we need validation logic hat this is not set when the OS is Windows

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3623,15 +3666,26 @@ type PodSecurityContext struct {
+	// TODO: update the default value to "Merge" when spec.os.name is not windows in v1.34
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
+	SupplementalGroupsPolicy *SupplementalGroupsPolicy
```
I don't see validation code for this?

thockin reviewed (commented):
API LGTM, just a couple small detals.

everpeace committed:
Added validations for SupplementalGroupsPolicy and ContainerUser

everpeace reviewed (commented):
Sorry for the lack of validation logic. Addressed. PTAL 🙇

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4318,6 +4318,11 @@ func validateWindows(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
 		if securityContext.SupplementalGroups != nil {
 			allErrs = append(allErrs, field.Forbidden(fldPath.Child("securityContext").Child("supplementalGroups"), "cannot be set for a windows pod"))
 		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#using-a-feature-gate-in-code

Dont need to check gates here.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4948,6 +4953,12 @@ func validatePodSpecSecurityContext(securityContext *core.PodSecurityContext, sp
 		allErrs = append(allErrs, validateWindowsSecurityContextOptions(securityContext.WindowsOptions, fldPath.Child("windowsOptions"))...)
 		allErrs = append(allErrs, ValidateAppArmorProfileField(securityContext.AppArmorProfile, fldPath.Child("appArmorProfile"))...)
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
Same

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8121,3 +8138,56 @@ func validateNodeSelectorTermHasOnlyAdditions(newTerm, oldTerm core.NodeSelector
+func validateContainerStatusUsers(containerStatuses []core.ContainerStatus, fldPath *field.Path, podOS *core.PodOS) field.ErrorList {
+	allErrors := field.ErrorList{}
+	osName := core.Linux
+	if podOS != nil {
```
I'm not super familiar with `core.PodOS` - when is it set?  What happens if it is not set?  Can we get here with this field being nil?  I don't see where scheduler considers it, and I don't see any default value defined.

@dchen1107

mrunalp replied:
It looks like this was introduced to detect Windows in admission plugins. https://github.com/kubernetes/enhancements/pull/2803/files
Scheduler interaction was marked as out of scope in that KEP.

thockin replied:
So this field is truly optional and advisory.  OK.

everpeace committed:
No need featuregate check in validation when adding new field with no default value

everpeace reviewed (commented):
Addressed comments for no need featuregate checks in validation logics. And rebased to the latest master.

thockin commented:
/approve for API

Needs sig-node review, please

mrunalp commented:
I will review for sig-node.

mrunalp commented on api/openapi-spec/v3/api__v1_openapi.json:
```diff
@@ -1595,6 +1595,14 @@
+                "$ref": "#/components/schemas/io.k8s.api.core.v1.ContainerUser"
+              }
+            ],
+            "description": "User represents user identitiy information of the first process in the container"
```
typo: identity

everpeace replied:
addressed in [5f0d761](https://github.com/kubernetes/kubernetes/pull/117842/commits/5f0d7613520d87745a229f8cc854bbf0e2e18221)

mrunalp commented on pkg/apis/core/types.go:
```diff
@@ -2741,6 +2741,31 @@ type ContainerStatus struct {
 	// +optional
 	// +featureGate=RecursiveReadOnlyMounts
 	VolumeMounts []VolumeMountStatus
+	// User represents user identitiy information of the first process in the container
```
typo: identity

everpeace replied:
addressed in [5f0d761](https://github.com/kubernetes/kubernetes/pull/117842/commits/5f0d7613520d87745a229f8cc854bbf0e2e18221)

mrunalp commented:
The node bits look fine. Only other nit I have is making the commits additive vs. fixing up earlier commits.
We can add e2es for this once the runtimes support this. Thanks!

everpeace committed:
fix typo: identitiy -> identity

everpeace commented:
> Only other nit I have is making the commits additive vs. fixing up earlier commits.

I think squashing my commits into one would be nice. Is it ok to squash?? 

> We can add e2es for this once the runtimes support this. Thanks!

For container runtimes supporting this feature, we need to merge this PR first as defined in [CRI API | Feature development](https://github.com/kubernetes/kubernetes/blob/6546c8b/staging/src/k8s.io/cri-api/README.md#feature-development). Thus, e2es will be added in another PR after containerd released the newer version which supports this. Am I right?

everpeace commented:
/test pull-kubernetes-node-e2e-containerd

mrunalp commented:
> Thus, e2es will be added in another PR after containerd released the newer version which supports this. Am I right?
Yes, that's right.

> I think squashing my commits into one would be nice. Are we ready to squash now??
The PR looks ready to me.

thockin commented:
/label tide/merge-method-squash

everpeace commented:
@thockin @mrunalp Thank you very very much for your review and patient nurturing🙇!

thockin commented:
No, thank you for your patience and perseverance!


Message ID: ***@***.***>
>
